File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*; 
class Solution {
    class Process{
        int pid; 
        int priority; 
        public Process(int pid, int priority){
            this.pid = pid; 
            this.priority = priority; 
        }
    }
    public int solution(int[] priorities, int location) {
        int answer = 0;
        /*
        큐 안에 있는 프로세스 중에서, 우선순위의 max값을 항상 알아야겠는데
        큐: 프로세스의 초기 인덱스 숫자, 내 우선순위 
        */
        ArrayDeque<int[]> q = new ArrayDeque<>(); 
        List<Integer> pList = new LinkedList<>();
        
        int seq = 1; 
        int maxP = -1; 
        for(int i=0; i<priorities.length; i++){
            q.add(new int[]{i, priorities[i]});
            pList.add(priorities[i]);
        }
        // pList = new ArrayList(new HashSet<>(pList)); 
        Collections.sort(pList, Collections.reverseOrder());
        /*
        0. 가장 큰 우선순위 구하기 -> maxP에 저장 
        1. 맨 앞에 있는 애가 locaiton == i인지 검사 
            맞으면 break seq가 answer임 
            아니면 2. 
        2. location == i가 될 때까지 while문 돌려 
            찾으면 stop 하고 break 
        */
        
        while(!q.isEmpty()){
           maxP = pList.remove(0);

           int[] cur = null; 
           while(true){
               cur = q.poll(); 
               if(maxP == cur[1]){
                   break; 
               }else{
                   q.offer(cur);
               }
           }

           if(location == cur[0]){
               answer = seq; 
               break; 
           }
           seq++; 
        }
        return answer;
    }
}

File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <string>
#include <vector>
#include <algorithm> // max(), min() 함수 있음
#include <queue> 
using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    // 프로세스번호, 우선순위 값 
    queue<pair<int, int>> q;
    // 우선순위 값을 내림차순으로 갖고 있음 
    // pq는 기본적으로 내림차순 정렬; 최대 힙임 
    priority_queue<int> pq; 
    
    for(int i=0; i<priorities.size(); i++){
        q.push({i, priorities[i]});
        pq.push(priorities[i]);
    }
    
    /**
    맨 앞에 있는 애를 뽑고 
    얘의 우선순위 == maxP -> pop -> maxP 갱신 
    더 작으면 -> pop -> push
    */
    int seq = 1; 
    while(!q.empty()){
        pair<int, int> cur = q.front(); 
        q.pop(); 
        
        int curNo = cur.first; 
        int curP = cur.second; 
        int maxP = pq.top(); 
        
        if(curP == maxP){
            // maxP 갱신 
            pq.pop(); 
            maxP = pq.top(); 
            // 찾는 프로세스면 겟 
            if(curNo == location) return seq; 
            seq++; 
        }else{
            // push 
            q.push(cur);
        }
    }
    return answer;
}
/*
프로세스가 몇 번째로 실행되는지 알아내기

전체적으로 maxPriority 값을 유지해야 할 것 같은데? 

가장 큰 우선순위 값을 만나면 pop 
아니면 push 

pop했으면 우선순위 값도 그 다음으로 큰 애로 변경

우선순위 값: 내림차순 정렬 
*/